client/client.go

```diff
@@ -171,15 +171,22 @@ type AllocRunner interface {
 // are expected to register as a schedule-able node to the servers, and to
 // run allocations as determined by the servers.
 type Client struct {
-	config *config.Config
-	start  time.Time
+	start time.Time
 
 	// stateDB is used to efficiently store client state.
 	stateDB state.StateDB
 
-	// configCopy is a copy that should be passed to alloc-runners.
-	configCopy *config.Config
-	configLock sync.RWMutex
+	// config must only be accessed with lock held. To update the config, use the
+	// Client.UpdateConfig() helper. If you need more fine grained control use
+	// the following pattern:
+	//
+	// 	c.configLock.Lock()
+	// 	newConfig := c.config.Copy()
+	// 	// <mutate newConfig>
+	// 	c.config = newConfig
+	// 	c.configLock.Unlock()
+	config     *config.Config
+	configLock sync.Mutex
```

Comment:
Yes! I only very briefly explored this route because the changes it entails get huge very quickly.

I think there's a middle ground that's the best of both worlds though: there may be so few mutable fields on `config.Config` that we could extract only them into getters/interfaces and let `config.Config` be immutable with no locking on access! `TLSConfig` already does this around key loading, although it's unfortunately jarring because it's the _only_ example of this pattern and the surrounding code can be difficult to trace.

Without checking `Node` and `TLSConfig` might be the only two mutable fields! I'm hoping to explore this with my dynamic node metadata work.